test/server/http/admin_test.cc

```diff
@@ -654,6 +654,23 @@ TEST_P(AdminInstanceTest, AdminProfiler) {
   EXPECT_FALSE(Profiler::Cpu::profilerEnabled());
 }
 
+TEST_P(AdminInstanceTest, AdminHeapProfiler) {
+  Buffer::OwnedImpl data;
+  Http::HeaderMapImpl header_map;
+
+#ifdef PROFILER_AVAILABLE
+  EXPECT_EQ(Http::Code::OK, postCallback("/heapprofiler?enable=y", header_map, data));
+  EXPECT_TRUE(Profiler::Heap::profilerEnabled());
```

Comment:
also test the enabling when it's already enabled, disabling when enabled, and disabling when already disabled.

Reply:
IHMO, it's not a huge issue. People will always have diverge opionions on whether an "enable" after "enable" should reset the profiling start point.
 
Could we just assume it's an undefined behavior as long as below hold?
- envoy is not crashing or leaking
- the last "disable" should dump and clean up the in memory resource

Reply:
Why don't we just define the behavior? You are only going to implement one behavior, so you might as well test it.

I'd vote for tenable-after-enable not doing a reset.  I think it should give a message saying "already enabled".  Maybe enable-disable-enable should reset it, or maybe there should be an explicit query-param to reset it. I'm not 100% of the semantics there.

But whatever you implement should be tested, and the resulting behavior reflected in the text output from the admin command.

Reply:
I personally also prefer already enabled as I implemented in the latest commit.
I have the new test that repeated requests wont crashing.

However, even with the implementation above I don't think anyone should rely on this behavior. As profiler user should understand when it is started. What's more, currently it's user's responsibility to guarantee no other malicious user secretly starts profiler or stops the profiler.

Reply:
I think you mean to test `Profiler::Heap::isProfilerStarted()` in this assertion (and below). AFAICT `profilerEnabled` tests whether the profiling is compiled into the build or not.

Reply:
You are right. I will fix.